internal/engine/fswatch/gitmanager.go

```diff
@@ -0,0 +1,138 @@
+package fswatch
+
+import (
+	"context"
+	"io/ioutil"
+	"path/filepath"
+	"time"
+
+	"github.com/tilt-dev/tilt/internal/store"
+	"github.com/tilt-dev/tilt/internal/watch"
+	"github.com/tilt-dev/tilt/pkg/logger"
+	"github.com/tilt-dev/tilt/pkg/model"
+)
+
+type repoNotifyCancel struct {
+	repo   model.LocalGitRepo
+	notify watch.Notify
+	cancel func()
+}
+
+// Watches git for branch switches and fires events.
+type GitManager struct {
+	fsWatcherMaker FsWatcherMaker
+	repoWatches    map[model.LocalGitRepo]repoNotifyCancel
+}
+
+func NewGitManager(fsWatcherMaker FsWatcherMaker) *GitManager {
+	return &GitManager{
+		fsWatcherMaker: fsWatcherMaker,
+		repoWatches:    make(map[model.LocalGitRepo]repoNotifyCancel),
+	}
+}
+
+func (m *GitManager) diff(ctx context.Context, st store.RStore) (setup, teardown []model.LocalGitRepo) {
+	state := st.RLockState()
+	defer st.RUnlockState()
+
+	if !state.EngineMode.WatchesFiles() {
+		return nil, nil
+	}
+
+	watchable := WatchableTargetsForManifests(state.Manifests())
+	reposToProcess := make(map[model.LocalGitRepo]bool)
+	for _, w := range watchable {
+		for _, repo := range w.LocalRepos() {
+			reposToProcess[repo] = true
+		}
+	}
+
+	for repo := range reposToProcess {
+		if _, ok := m.repoWatches[repo]; !ok {
+			setup = append(setup, repo)
+		}
+	}
+
+	for repo := range m.repoWatches {
+		if _, ok := reposToProcess[repo]; !ok {
+			teardown = append(teardown, repo)
+		}
+	}
+
+	return setup, teardown
+}
+
+func (m *GitManager) OnChange(ctx context.Context, st store.RStore) {
+	setup, teardown := m.diff(ctx, st)
+
+	for _, repo := range teardown {
+		p, ok := m.repoWatches[repo]
+		if !ok || p.notify == nil {
+			continue
```

Comment:
not a big deal, but if this the repo exists in `m.repoWatches` w/o being successfully started, this means we'll never REMOVE it from `m.repoWatches` and it'll hit this teardown loop every time (or else if we ever WANT to watch this repo again, we won't attempt to START the watch again b/c it's already in `repoWatches`), right?